test/test_before_all.py

```diff
@@ -1,9 +1,9 @@
-import pytest
 import subprocess
 import textwrap
 
-from . import utils
-from . import test_projects
+import pytest
+
+from . import test_projects, utils
```

Comment:
This is slightly confusing, actually, the thing after a `from x import` ideally shouldn't be a module; if you have an `__init__.py` and it contains a `test_projects` or `utils` object, you'll get that instead of the module. But it's a bit special for `.` imports, since you are required to use `from` with a relative import.

I believe you are guaranteed to have the test directory as the first on the path in PyTest, so this could be just `import test_projects`, etc? Also, we could `from .test_projects import new_c_project`, etc.

Reply:
Some of the exiting ones are written this way, so I'd probably leave it for now, and eventually rewrite them properly (see `test_0_basic.py`, for example).

Reply:
Yeah, I first wanted to suggest `import .test_projects`, but apparently that only works in the `from ... import ...` construct

Reply:
> the thing after a `from x import` ideally shouldn't be a module

Not sure I agree, e.g. `from numpy import linalg` `from django import models` `from tensorflow import keras`... also I'd rather rely on the `__init__.py` file being empty than rely on the value of `sys.path`. I expect tools like mypy or language servers to be able to read the code better without relying on a `sys.path` set by pytest too.

Reply:
`from numpy import linalg` is an often used shortcut for `import numpy.linalg as linalg`, but it is confusing - a reader can't tell if linalg is a package or an object. We are stuck with it, but I wish the language wasn't ambiguous. Ideally, relative imports probably should import objects, `from .test_projects import new_c_project`.

Relative imports also don't work if you don't execute it as a package; `python test_before_all.py` will fail. (`ImportError: attempted relative import with no known parent package`) (PyTest handles this correctly).

Reply:
A "tests" directory (or worse, "test") isn't really a package - `import test` imports the stdlib `test`. I think PyTest does some shenanigans to make it work.

Reply:
We are off on a big ol' tangent here, but-

> a reader can't tell if linalg is a package or an object

I'm not sure it matters, from a users' point of view. It's still an imported namespace, or global state if you mutate it. Probably I've just used different bits of Python to you, but to me, importing a module like this is pretty common.

Reply:
Further aside - I often see people talk negatively of relative imports, in general... why is that? I've never had trouble understanding them, and they seem explicit enough ¯\_(ツ)_/¯

Reply:
> I'm not sure it matters, from a users' point of view.

If you try to rewrite it, you have to either run it interactively or look up the source. For example:

```python
from a import b
b.c()
# May or may not be written as
from a.b import c
c()
```

And, if you need to look up the source, is there an `a/b.py` or an `a/__init__.py` file with a b in it, or is it a file? (I often teach Python...)

Reply:
I use relative imports a lot, but they have downsides, especially for new users who do not know how to set up a proper environment. They _have_ to be in a loaded package; it's not only "in the same directory". So `python x.py` inside a package fails if there are relative imports in `x`; I think `python outer/x.py` fails too (I could be wrong). If you stick a `if __name__ == "__main__"` in it for standalone runs or tests, it fails.

Reply:
I also use absolute imports a lot...

Reply:
I do know someone who will not do any import except `import a.b.c`, and always uses fully qualified names. That's it. No shortening. It drive me crazy seeing `numpy.stuff` sometimes. :)

Reply:
FWIW (given we're on this tangent anyway?), I tend to not like `from module import thing` either. Often, `module.thing` is not so much longer to write, should be auto-completed, and we're not trying to save bytes in code anyway? But yes, `numpy` to `np` just becomes a habit, after a while :-)